ssl/ssl_key_share.cc

```diff
@@ -176,20 +178,550 @@ class X25519KeyShare : public SSLKeyShare {
+    // peer by writing it to |out_public_key|, then...
+    Array<uint8_t> ciphertext;
+    Array<uint8_t> shared_secret;
+    if (!ciphertext.Init(ciphertext_len) || !shared_secret.Init(secret_len)) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_MALLOC_FAILURE);
+      return false;
+    }
+
+    size_t ciphertext_bytes_written = ciphertext_len;
+    size_t secret_bytes_written = secret_len;
+    if (!EVP_PKEY_encapsulate(ctx_.get(), ciphertext.data(),
+                              &ciphertext_bytes_written, shared_secret.data(),
+                              &secret_bytes_written) ||
+                              ciphertext_bytes_written != ciphertext_len ||
+                              secret_bytes_written != secret_len ||
+                              !CBB_add_bytes(out_public_key, ciphertext.data(), ciphertext_len)) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_INTERNAL_ERROR);
+      return false;
+    }
+
+    // ...we retain the shared secret for our own use.
+    *out_secret = std::move(shared_secret);
+
+    // Success; clear the alert
+    *out_alert = 0;
+    return true;
+  }
+
+  // Because this is a KEM key share, |peer_key| is actually the ciphertext
+  // resulting from the peer encapsulating the shared secret under our public key.
+  // In Finish(), we use our previously generated private key to decrypt
+  // that ciphertext and obtain the shared secret.
+  bool Finish(Array<uint8_t> *out_secret, uint8_t *out_alert,
+              Span<const uint8_t> peer_key) override {
+    // Basic nullptr checks
+    if (!out_alert) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_PASSED_NULL_PARAMETER);
+      return false;
+    }
+
+    // Set alert to internal error by default
+    *out_alert = SSL_AD_INTERNAL_ERROR;
+
+    if (!out_secret) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_PASSED_NULL_PARAMETER);
+      return false;
+    }
+
+    // ctx_ should have been initialized by a previous call to Offer()
+    if (!ctx_) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_SHOULD_NOT_HAVE_BEEN_CALLED);
+      return false;
+    }
+
+    // Retrieve the lengths of the ciphertext and shared secret
+    size_t ciphertext_len = 0;
+    size_t secret_len = 0;
+    if (!EVP_PKEY_encapsulate(ctx_.get(), nullptr, &ciphertext_len, nullptr,
+                              &secret_len)) {
+      OPENSSL_PUT_ERROR(SSL, ERR_R_INTERNAL_ERROR);
+      return false;
+    }
+
+    // Ensure that peer_key is valid
+    if (!peer_key.data() || peer_key.size() != ciphertext_len) {
+      *out_alert = SSL_AD_DECODE_ERROR;
+      OPENSSL_PUT_ERROR(SSL, ERR_R_INTERNAL_ERROR);
+      return false;
+    }
+
+    // |peer_key| is the ciphertext, encapsulating the shared secret,
+    // that the peer generated using our public key. We use our private
+    // key to decapsulate it, and retain the shared secret by writing
+    // it to |out_secret|.
+    uint8_t *ciphertext = (uint8_t *)peer_key.begin();
+    if (!out_secret->Init(secret_len)) {
```

Comment:
Any reason you initialized out_secret here and wrote to it directly instead of moving the final result like other places above?

Reply:
I believe the other places that perform the `std::move()` pattern to ensure that in the event of any of the function calls between `shared_secret.Init()` and the Encapsulate/Decapsulate failing (such as CBB memory allocations), that `out_secret` isn't modified. Here, there aren't multiple function calls between those two points, so the move pattern isn't used.